python/ray/exceptions.py

```diff
@@ -0,0 +1,97 @@
+import os
+import colorama
+
+try:
+    import setproctitle
+except ImportError:
+    setproctitle = None
+
+
+class RayError(Exception):
+    """Super class of all ray exception types."""
+    pass
+
+
+class RayTaskError(RayError):
+    """Indicates that a task threw an exception during execution.
+
+    If a task throws an exception during execution, a RayTaskError is stored in
+    the object store for each of the task's outputs. When an object is
+    retrieved from the object store, the Python method that retrieved it checks
+    to see if the object is a RayTaskError and if it is then an exception is
+    thrown propagating the error message.
+
+    Attributes:
+        function_name (str): The name of the function that failed and produced
+            the RayTaskError.
+        traceback_str (str): The traceback from the exception.
+    """
+
+    def __init__(self, function_name, traceback_str):
+        """Initialize a RayTaskError."""
+        if setproctitle:
+            self.proctitle = setproctitle.getproctitle()
+        else:
+            self.proctitle = "ray_worker"
+        self.pid = os.getpid()
+        self.host = os.uname()[1]
+        self.function_name = function_name
+        self.traceback_str = traceback_str
+        assert traceback_str is not None
+
+    def __str__(self):
+        """Format a RayTaskError as a string."""
+        lines = self.traceback_str.split("\n")
+        out = []
+        in_worker = False
+        for line in lines:
+            if line.startswith("Traceback "):
+                out.append("{}{}{} (pid={}, host={})".format(
+                    colorama.Fore.CYAN, self.proctitle, colorama.Fore.RESET,
+                    self.pid, self.host))
+            elif in_worker:
+                in_worker = False
+            elif "ray/worker.py" in line or "ray/function_manager.py" in line:
+                in_worker = True
+            else:
+                out.append(line)
+        return "\n".join(out)
+
+
+class RayWorkerError(RayError):
+    """Indicates that the worker died unexpectedly while executing a task."""
+
+    def __str__(self):
+        return "The worker died unexpectedly while executing this task."
+
+
+class RayActorError(RayError):
+    """Indicates that the actor died unexpectedly before finish executing a
+    task."""
+
+    def __init__(self):
+        return "The actor died unexpectedly before finish executing this task."
+
+
+class UnreconstructableError(RayError):
```

Comment:
I would maybe call this `RayObjectEvictedError` since eviction is the most common cause. Or at least modify the message to include something about this being an evicted actor object.

Reply:
I slightly prefer keeping this `UnreconstructableError`. Because object eviction doesn't always lead to errors. If the evicted object is from a normal task, it can always be reconstructed; if it's from an actor and the actor is restored from a previous checkpoint, the object can be reconstructed as well. Only when we try to get an unreconstructable object, an error will occur. 
But I'll add more explanation about evicted actor object.